src/btree.h

```diff
@@ -0,0 +1,81 @@
+#include "buffer.h"
+#include <stdexcept>
+#include <stddef.h>
+
+template <typename K>
+class rootNode {
+
+};
+
+template <typename K>
+class leafNode {
+
+};
+
+template <typename K>
+class internalNode {
+
+};
+
+/**
+	Some of the fields available in header of BufferedFile
+
+ 	const size_type key_size;
+	const size_type value_size;
+	const string Identifier = "BTREE";
+
+	These can be accessed as :
+	$buffered_file_object->header->key_size;
+
+ */
+
+
+template <typename K, typename V>
+class BTree
+{
+public:
+	typedef long long int size_type;
+private:
+	BufferedFile* buffered_file_internal;
+	BufferedFile* buffered_file_data;
+
+	rootNode *root;
+	size_type sz;
+	long M; // Maximum M keys in the internal nodes
+
+public:
+	BTree(const char* pathname, size_type blocksize) : block_size(blocksize), element_size(sizeof(T)),
+		sz(0)
+	) {
+
+		// this would have read the header from the file and set all the fields accordingly
+		buffered_file_internal = new BufferedFile(pathname, block_size, block_size*3);
+
+		// get data file
+		buffered_file_data = new BufferedFile(pathname + "_data", sizeof(V), sizeof(V)*3);
+
+		// will write it to the appropriate position in the header
+		buffered_file_internal->header->setDataFileDescr(
+			buffered_file_data->getFileDescriptor()
+		);
+
+		root = NULL;
+
+		// now ready for operation
+	}
+
+	~BTree() {
+		// save the header to the file on disk
+		buffered_file_internal->writeHeaderPersistent();
```

Comment:
if we have made any changes to the header in memory i.e. `buffered_file_internal->header`, before deleting it from the memory, we write the header to the file and then delete it.